src/runtime/hexagon_dma_device_interface.cpp

```diff
@@ -0,0 +1,251 @@
+            return HEX_ERROR;
+        }
+
+        if (halide_hexagon_dmart_set_dma_handle(user_context, pdma_context, handle, frame)) {
+            error(user_context) << "halide_hexagon_dma_device_malloc:Function failed to set DMA Handle to DMA context \n";
+            return HEX_ERROR;
+        }
+    }
+    return 0;
+}
+
+WEAK int halide_hexagon_dma_device_free(void *user_context, halide_buffer_t *buf) {
+    halide_assert(NULL, buf!=NULL);
+    HexagonDmaContext hexagon_dma_context(user_context);
+    halide_assert(user_context, hexagon_dma_context.get_context() != NULL);
+    p_dma_context dma_ctxt = hexagon_dma_context.get_context();
+
+    halide_hexagon_dmart_delete_context(user_context, dma_ctxt);
+    return HEX_SUCCESS;
+}
+
+WEAK int halide_hexagon_dma_copy_to_device(void *user_context,  halide_buffer_t *buf) {
+    void* handle;
+
+    //////////////////////////////////////////////////////////////////////
+    p_dma_context dma_handle = reinterpret_cast<p_dma_context>(buf->device);
+    uintptr_t  frame = halide_hexagon_dmart_get_frame(user_context, dma_handle);
+    //////////////////////////////////////////////////////////////////////////
+
+    handle = halide_hexagon_dmart_get_dma_handle(user_context, dma_handle, frame);
+    if (handle == 0) {
+        error(user_context) << "Function failed to get DMA Write  Handle  \n";
+        return HEX_ERROR;
+    }
+
+    int ncomponents;
+    ncomponents = halide_hexagon_dmart_get_num_components(user_context, dma_handle, frame);
+
+    t_dma_move_params move_param;
+
+    uintptr_t fold_addr = halide_hexagon_dmart_get_fold_addr(user_context, dma_handle, frame);
+    halide_hexagon_dmart_get_update_params(user_context, dma_handle, fold_addr, &move_param);
+
+    move_param.handle = handle;
+    move_param.ncomponents = ncomponents;
+
+    int nRet = HEX_SUCCESS;
+    nRet = dma_move_data(move_param);
+
+    if (nRet != QURT_EOK) {
+        return HEX_ERROR;
+    }
+    return HEX_SUCCESS;
+}
+
+WEAK int halide_hexagon_dma_copy_to_host(void *user_context, halide_buffer_t *buf) {
+    void* handle;
+
+    //////////////////////////////////////////////////////////////////////
+    p_dma_context dma_handle = reinterpret_cast<p_dma_context>(buf->device);
+    uintptr_t frame = halide_hexagon_dmart_get_frame(user_context, dma_handle);
+    //////////////////////////////////////////////////////////////////////////
+
+    handle = halide_hexagon_dmart_get_dma_handle(user_context, dma_handle, frame);
+    if (handle == 0) {
+        error(user_context) << "Function failed to get DMA Write  Handle  \n";
+        return HEX_ERROR;
+    }
+
+    int ncomponents;
+    ncomponents = halide_hexagon_dmart_get_num_components(user_context, dma_handle, frame);
+
+    t_dma_move_params move_param;
+
+    uintptr_t fold_addr = halide_hexagon_dmart_get_fold_addr(user_context, dma_handle, frame);
```

Comment:
Is fold_addr going to be the host pointer of buf? It needs to be, or this function isn't doing what it's supposed to do.